### ISSUE TITLE
[Bug] Qwen3.5-35B-A3B stops abruptly during token generation (multi-turn with Claude / OpenCode)

### ISSUE THREAD

**Issue by c4software** [jundot/omlx#260]
### Describe the bug
Using oMLX (latest version), the **Qwen3.5-35B-A3B** model.

The issue occurs especially in **multi-turn conversations** (long coding sessions typical of agents like Claude Code or OpenCode). No visible error in the logs, the conversation is just ended. I need to say continue to restart the token generation.

### To Reproduce
1. Launch oMLX (menu bar app or CLI).
2. Load the model.
3. Connect via the OpenAI-compatible endpoint (`http://localhost:8000/v1`) using Claude or OpenCode.
4. Have a 2-3 turn conversation / tools usage (especially long reasoning/coding).

### Expected behavior
Smooth, continuous generation even on long multi-turn responses, thanks to oMLX’s tiered KV cache SSD.

### Environment
- oMLX version: Latest
- Model: Qwen3.5-35B-A3B-4bit
- Mac: M3 Pro - 36Gb
- Client: Claude Code / OpenCode
- Quantization: 4-bit

Thanks for the amazing tool (the tiered KV cache is awesome for coding agents)!

**Comment by c4software** [comment]
For example :

<img width="1185" height="814" alt="Image" src="https://github.com/user-attachments/assets/cde575d0-ab05-470f-93a1-cf0f840e79dc" />

**Comment by jundot** [comment]
This issue has been reported and discussed multiple times in #205, and several patches have been applied to address it. The core fix should already be included in recent versions.

In some cases though, this is actually a model limitation - certain models struggle to produce proper tool calling output in long context conversations, and that can look identical to this bug from the client side.

To figure out what's going on, i need to see the actual raw model response. Could you share a trace-level server log?

1. Clear the existing log (delete `server.log` or find it at `~/Library/Application Support/oMLX/logs/server.log`)
2. Open the web dashboard → global settings → set log level to **trace** (this records the full raw model response)
3. Reproduce the issue
4. Share the `server.log` file

If your conversation contains anything sensitive (code, credentials, etc.), feel free to redact those parts - i just need the model's raw output around where it stops.

**Comment by c4software** [comment]
Thanks for the quick feedback. I will capture the log tomorrow. Meanwhile, I think I found a possible cause of the problem, if I disable the cache the problem seems not present.

**Comment by jundot** [comment]
@c4software

Your observation about disabling cache was spot on. I dug into it and found what was going on.

Qwen3.5's GatedDeltaNet layers maintain a recurrent state that accumulates through the sequence. During prefill the prompt gets processed in chunks, and between each chunk the state goes through a float32 to bfloat16 roundtrip. The boundary cache block size was 1024 while prefill step size was 2048, so with cache enabled the model was doing twice as many roundtrips. That accumulated precision loss was enough to shift logits and make the model output EOS instead of tool calls.

Fixed in v0.2.15b1 by aligning the block size to match prefill step size so the chunk count is identical regardless of cache being on or off. Also patched a missing `cache.advance()` call in mlx-lm's qwen3_5 that could cause issues when batching multiple requests.

Tested across 4 model types (KVCache, ArraysCache, RotatingKVCache) at temp=0 with 8K context. Boundary ON/OFF and SSD cache hit/fresh all produce identical tokens on ArraysCache models.

- macOS 15 Sequoia: https://omlx.ai/beta/oMLX-0.2.15b1-macos15-sequoia_260317112406.dmg
- macOS 26 Tahoe: https://omlx.ai/beta/oMLX-0.2.15b1-macos26-tahoe_260317112622.dmg

This is a beta build. If testing goes well i'll include it in the next stable release. Would appreciate if you could give it a try and let me know how it goes.

**Comment by jundot** [comment]
@c4software

After patching i ran more tests with Qwen3.5-35B-A3B-8bit and the tool calling still breaks in long context sessions. So i wrote a separate test script that compares token-level output between pure mlx-lm inference (no omlx, no cache) and omlx with boundary cache ON/OFF at temperature 0.

All three produce identical tokens:

```
Prompt: 4114 tokens, generating 100 tokens

Results:
  pure mlx-lm vs boundary cache OFF: IDENTICAL
  pure mlx-lm vs boundary cache ON:  IDENTICAL
  boundary cache OFF vs ON:          IDENTICAL
```

This confirms omlx's caching layer is not altering the model output at all. The tool calling breakage in long context is coming from the model itself. Qwen3.5-35B-A3B has only 3B active parameters (hidden_size=2048), and it seems to lose coherence for tool calling beyond a certain context length. Even at 8bit quantization (which should preserve most of the model's capability), the same behavior shows up, which points more toward a model architecture limitation than a quantization accuracy issue.

The boundary cache fix in v0.2.15b1 is still worth applying since it eliminates a real source of numerical divergence between cache ON/OFF. But it won't fix cases where the model itself stops producing tool calls at long context.

**Comment by jundot** [comment]
fixed in [v0.2.15](https://github.com/jundot/omlx/releases/tag/v0.2.15)

**Comment by viljark** [comment]
I am using Qwen3.5-35B-A3B-4bit and I'm still seeing similar behaviour in **v0.2.17,** the generated text is just cut off without even ending the `<think>` tag. Happens with both caching enabled and disabled. 
NB I have an image in the first message.

I'm out of ideas, maybe there is something I could try more or I'm reaching model's limits at this point where it generates stop tokens without completing the reasoning?

```
2026-03-17 14:47:56,533 - omlx.server - TRACE - [-] -   Message[27]: role=user, content=[ContentPart(type='text', text='use git status', image_url=None)]...
2026-03-17 14:47:56,533 - omlx.server - TRACE - [-] -   Message[28]: role=assistant, content=(empty)...
2026-03-17 14:47:56,533 - omlx.server - TRACE - [-] -   Message[29]: role=tool, content=On branch feature/projects-list-progress-column
Your branch is up to date with 'origin/feature/projects-list-progress-column'.

Changes not staged for commit:
  (use "git add <file>..." to update what...
2026-03-17 14:47:56,534 - PIL.PngImagePlugin - DEBUG - [-] - STREAM b'IHDR' 16 13
2026-03-17 14:47:56,534 - PIL.PngImagePlugin - DEBUG - [-] - STREAM b'IDAT' 41 50609
2026-03-17 14:47:56,552 - omlx.server - DEBUG - [-] - Sampling params: temperature=0.6, top_p=0.95, top_k=20, repetition_penalty=1.0, min_p=0.0, presence_penalty=0.0, frequency_penalty=0.0, max_tokens=32768 (model: Qwen3.5-35B-A3B-4bit)
2026-03-17 14:47:56,552 - PIL.PngImagePlugin - DEBUG - [-] - STREAM b'IHDR' 16 13
2026-03-17 14:47:56,552 - PIL.PngImagePlugin - DEBUG - [-] - STREAM b'IDAT' 41 50609
2026-03-17 14:47:56,748 - omlx.scheduler - DEBUG - [-] - Added request 3221f0c1-bdfd-44dd-866a-0c5f1266f3f8 with 2548 prompt tokens
2026-03-17 14:47:56,748 - omlx.scheduler - DEBUG - [-] - Scheduled request 3221f0c1-bdfd-44dd-866a-0c5f1266f3f8 (uid=1) with 2548 tokens to process (2548 total), no cache
2026-03-17 14:47:59,231 - omlx.scheduler - DEBUG - [-] - Request 3221f0c1-bdfd-44dd-866a-0c5f1266f3f8 finished: stop, 18 tokens
2026-03-17 14:47:59,231 - omlx.scheduler - TRACE - [-] - Request 3221f0c1-bdfd-44dd-866a-0c5f1266f3f8 generated text:
<think>
The file path in the CR feedback seems incorrect. Let me search for the actual file:
2026-03-17 14:48:00,649 - omlx.server - WARNING - [-] - GET /admin/api/stats → 401: Admin authentication required
```

Session:
 Tokens
 Input: 28 591
 Output: 733
 Total: 29 324

Settings:
Max Context Window 64000
Max Tokens 32768

-update-
When I switched the model to Qwen3-Coder-30B-A3B-Instruct-4bit-DWQ, then it was able to complete the task.

**Comment by Bradley-Liu** [comment]
@c4software are you able to run glm4.7 flash on your mac. i have mac with the same amount of vram as yours. when i use glm4.7 in claude code and openclaw. it frequently gave me this error "● [Error: Request aborted: process memory limit exceeded. Increase --max-process-memory or reduce   

  context size.]   " 

When i use qwen3.5 35b a3b. it won't although qwen3.5 is large than glm4.7 both 4bit.  Thank you. I also found that qwen3.5 is really weak when calling tools or skills in claude code. What is your experience.

**Comment by c4software** [comment]
> @c4software are you able to run glm4.7 flash on your mac. i have mac with the same amount of vram as yours. when i use glm4.7 in claude code and openclaw. it frequently gave me this error "● [Error: Request aborted: process memory limit exceeded. Increase --max-process-memory or reduce   
> 
>   context size.]   " 
> 
> When i use qwen3.5 35b a3b. it won't although qwen3.5 is large than glm4.7 both 4bit.  Thank you. I also found that qwen3.5 is really weak when calling tools or skills in claude code. What is your experience. 

I have the same issues. I also have this with really long exchange using qwen3.5 35b a3b. 

But with Qwen and tool calling, I don't have big issues with tool calling. But as usual, with local models, the results with large problems are often disappointing. But, for small code update or driven with a large model (like doing the plan with Sonnet, and work with Qwen)

**Comment by Bradley-Liu** [comment]
@c4software I think it has something to do with the glm4.7 model, not omlx. qwen3.5 35b a3b is the only model i can use without any problems. are there any other models that you recommend that can be used without problems in omlx?

**Comment by c4software** [comment]
> @c4software I think it has something to do with the glm4.7 model, not omlx. qwen3.5 35b a3b is the only model i can use without any problems. are there any other models that you recommend that can be used without problems in omlx?

I've only tested with the Qwen 3.5 series, but with pure MLX the GPT-OSS 20B is quite nice too.

**Comment by jundot** [comment]
I tested this with settings that produce token-level identical output to pure mlx-lm, and the tool calling issue still shows up in long contexts. This looks like a model-side problem rather than something on the oMLX side, so i'm closing this for now.

If you have anything else to discuss or new findings, feel free to reopen this or create a new issue anytime.
